File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        
        int n = nums2.size();
        int m = nums1.size();

        stack<int> monostack;
        vector<int> ans(n, -1);
        unordered_map<int, int> umap;

        for (int i = 0; i < n; i++) {

            while (!monostack.empty() && nums2[monostack.top()] < nums2[i]) {
                ans[monostack.top()] = nums2[i];
                monostack.pop();
            }

            monostack.push(i);
        }

        for (int i = 0; i < n; i++) {
            umap[nums2[i]] = ans[i];
        }

        vector<int> result;

        for (int i = 0; i < m; i++) {
            result.push_back(umap[nums1[i]]);
        }

        return result;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    # 4. Next Greater Element I (LeetCode 496)
    def nextGreaterElement(self, nums1: list[int], nums2: list[int]) -> list[int]:
        # Dictionary to store the next greater element for each number in nums2
        nge_map = {}
        # Monotonic decreasing stack to track elements waiting for a greater number
        stack = []
        
        for num in nums2:
            # While stack is not empty and current number is greater than the stack's top element
            while stack and num > stack[-1]:
                nge_map[stack.pop()] = num
            stack.append(num)
            
        # For any remaining elements in the stack, there is no next greater element (-1)
        while stack:
            nge_map[stack.pop()] = -1
            
        # Build the result for nums1 using the precomputed map
        return [nge_map[num] for num in nums1]